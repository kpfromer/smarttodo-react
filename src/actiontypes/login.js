export const INVALIDATE_LOGIN = 'login/INVALIDATE_LOGIN';
export const REQUEST_LOGIN = 'login/REQUEST_LOGIN';
export const RECEIVE_LOGIN = 'login/RECEIVE_LOGIN';
export const FAILURE_LOGIN = 'login/FAILURE_LOGIN';