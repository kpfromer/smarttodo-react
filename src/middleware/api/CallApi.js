import { API_ROOT } from "./api";
import axios from 'axios';

export default (endpoint, method, mapResponse = value => value, data = undefined, token = undefined) => {

  const fullUrl = endpoint.includes(API_ROOT) ? endpoint : API_ROOT + endpoint;

  return axios({
    url: fullUrl,
    method,
    headers: {
      'Content-Type': 'application/json',
      'Authorization': token ? `Bearer ${token}` : undefined
    },
    data
  })
    .then(response => mapResponse(response.data))
    .catch(error => Promise.reject(error));
};
