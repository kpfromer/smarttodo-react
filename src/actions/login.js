import * as LoginActionTypes from '../actiontypes/login';
import { CALL_API } from "../middleware/api/api";

export const tryLogin = (username, password) => ({
  [CALL_API]: {
    types: [
      LoginActionTypes.REQUEST_LOGIN,
      LoginActionTypes.RECEIVE_LOGIN,
      LoginActionTypes.FAILURE_LOGIN
    ],
    method: 'POST',
    endpoint: '/login',
    data: { username, password },
    authenticate: false,
    mapResponse: response => response.access_token
  }
});

export const logout = () => ({
  type: LoginActionTypes.INVALIDATE_LOGIN
});