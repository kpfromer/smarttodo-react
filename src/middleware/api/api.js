import CallApi from './CallApi';

export const API_ROOT = 'http://localhost:3001/v1';
export const CALL_API = 'call/CALL_API';

const CREATE_TYPES = [
  'POST',
  'PUT'
];
const VALID_TYPES = [
  ...CREATE_TYPES,
  'GET',
  'DELETE'
];

export default ({ getAuthFromState } = { getAuthFromState: val => val }) => store => next => action => {
  const callAPI = action[CALL_API];

  if (typeof callAPI === 'undefined') {
    return next(action);
  }

  let { endpoint } = callAPI;
  const { types, mapResponse, method, data, authenticate = true } = callAPI;

  if (typeof endpoint === 'function') {
    endpoint = endpoint(store.getState())
  }

  if (typeof endpoint !== 'string') {
    throw new Error('Specify a string endpoint URL.')
  }
  if (!Array.isArray(types) || types.length !== 3) {
    throw new Error('Expected an array of three action types.')
  }
  if (!types.every(type => typeof type === 'string')) {
    throw new Error('Expected action types to be strings.')
  }

  if (!VALID_TYPES.includes(method)) {
    throw new Error(`Invalid call api method: ${method}`);
  }

  if (CREATE_TYPES.includes(method) && !data) {
    throw new Error(`Method: ${method} requires data. Data is currently: ${data}`);
  }

  const actionWith = data => {
    const finalAction = { ...action, ...data };
    delete finalAction[CALL_API];
    return finalAction;
  };

  let token;

  if (authenticate) {
    token = getAuthFromState(store.getState());
  }

  const [requestType, successType, failureType] = types;
  next(actionWith({ type: requestType }));

  return CallApi(endpoint, method, mapResponse, data, token).then(
    response => next(actionWith({
      response,
      type: successType
    })),
    error => next(actionWith({
      type: failureType,
      error: error.message || 'Something bad happened'
    }))
  )
};
