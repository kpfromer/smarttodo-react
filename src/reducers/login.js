import * as LoginActionTypes from '../actiontypes/login';

function login
(
  state = {
    isFetching: false,
    didInvalidate: false,
    token: localStorage.getItem('id') || null
  },
  action
) {
  switch (action.type) {
    case LoginActionTypes.INVALIDATE_LOGIN:
      localStorage.removeItem('id');
      return {
        ...state,
        token: null,
        didInvalidate: true
      };
    case LoginActionTypes.REQUEST_LOGIN:
      return {
        ...state,
        isFetching: true,
        didInvalidate: false
      };
    case LoginActionTypes.RECEIVE_LOGIN:
      localStorage.setItem('id', action.response);
      return {
        ...state,
        isFetching: false,
        didInvalidate: false,
        token: action.response
      };
    case LoginActionTypes.FAILURE_LOGIN:
      return {
        ...state,
        isFetching: false
      };
    default:
      return state;
  }
}

export default login