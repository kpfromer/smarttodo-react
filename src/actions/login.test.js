import { logout, tryLogin } from "./login";
import { CALL_API } from "../middleware/api/api";

describe('login actions', () => {
  describe('tryLogin', () => {
    let username, password, action;

    beforeEach(() => {
      username = 'iAmARadUsername';
      password = 'verysecurepassword';
      action = tryLogin(username, password);
    });

    it('maps data to access token', () => {
      const data = {
        access_token: 'mockToken!'
      };
      expect(action[CALL_API].mapResponse(data)).toBe(data.access_token);
    });

    it('creates an api call action to login', () => {
      expect(action[CALL_API]).toMatchSnapshot();
    });
  });

  describe('logout', () => {
    let action;

    beforeEach(() => {
      action = logout();
    });

    it('creates an invalidate login action', () => {
      expect(action).toMatchSnapshot();
    });
  });
});