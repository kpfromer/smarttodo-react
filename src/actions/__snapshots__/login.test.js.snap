// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login actions logout creates an invalidate login action 1`] = `
Object {
  "type": "login/INVALIDATE_LOGIN",
}
`;

exports[`login actions tryLogin creates an api call action to login 1`] = `
Object {
  "authenticate": false,
  "data": Object {
    "password": "verysecurepassword",
    "username": "iAmARadUsername",
  },
  "endpoint": "/login",
  "mapResponse": [Function],
  "method": "POST",
  "types": Array [
    "login/REQUEST_LOGIN",
    "login/RECEIVE_LOGIN",
    "login/FAILURE_LOGIN",
  ],
}
`;
